### PR TITLE
Implement additional `SearchIssuesRequest` options

### LINKED ISSUE #1227: Missing options on `SearchIssuesRequest`
There are a number of "filter" fields/options mentioned in [the API documentation] and [how-to search article](https://help.github.com/articles/searching-issues) that are not currently implemented in SearchIssuesRequest class

- Searching by [`team`](https://help.github.com/articles/searching-issues/#search-by-a-team-thats-mentioned-within-an-issue-or-pull-request)
- Searching by "not" label syntax [`-label`](https://help.github.com/articles/searching-issues/#search-by-the-labels-on-an-issue)
- Searching on missing metadata [`no`](https://help.github.com/articles/searching-issues/#search-by-missing-metadata-on-an-issue-or-pull-request)
- The "human" styled [`is`](https://help.github.com/articles/searching-issues/#search-based-on-the-state-of-an-issue-or-pull-request) query 
 - (although `state` and `type` cover most of the `is` options, the `merged`/`unmerged` have no alternative)
- Searching by commit [`status`](https://help.github.com/articles/searching-issues/#search-based-on-commit-status)
- Searching by branch names [`base`](https://help.github.com/articles/searching-issues/#search-based-on-branch-names) or [`head`](https://help.github.com/articles/searching-issues/#search-based-on-branch-names)
- Searching on [`closed`](https://help.github.com/articles/searching-issues/#search-based-on-when-an-issue-or-pull-request-was-closed) date range

### PR DESCRIPTION
**Pull request by ryangribble**
## Status
- [x] Implemented new options
- [x] Unit tests for the request parameters for all new (and existing types)
- [x] Flesh out some integration tests to use some of the new options
- [x] Get feedback on naming of `No` property and `IssueNoMetadataQualifier` enum
- [x] Get feedback on implementation of exclusion parameters
- [x] Guidance from @Haacked +/- @shiftkey on whether to include additional fields other than `label` in the exclusion support

## Details
Fixes #1227 
https://developer.github.com/v3/search/#search-issues
https://help.github.com/articles/searching-issues

Added new options to `SearchIssuesRequest`

- Searching by [`team`](https://help.github.com/articles/searching-issues/#search-by-a-team-thats-mentioned-within-an-issue-or-pull-request)
- Searching by "not" label syntax [`-label`](https://help.github.com/articles/searching-issues/#search-by-the-labels-on-an-issue)
- Searching on missing metadata [`no`](https://help.github.com/articles/searching-issues/#search-by-missing-metadata-on-an-issue-or-pull-request)
- The "human" styled [`is`](https://help.github.com/articles/searching-issues/#search-based-on-the-state-of-an-issue-or-pull-request) query 
 - (although `state` and `type` cover most of the `is` options, the `merged`/`unmerged` have no alternative)
- Searching by commit [`status`](https://help.github.com/articles/searching-issues/#search-based-on-commit-status)
- Searching by branch names [`base`](https://help.github.com/articles/searching-issues/#search-based-on-branch-names) or [`head`](https://help.github.com/articles/searching-issues/#search-based-on-branch-names)
- Searching on [`closed`](https://help.github.com/articles/searching-issues/#search-based-on-when-an-issue-or-pull-request-was-closed) date range

## Testing
Added unit tests for each parameter on this class to verify that when not set, the parameter is not in the query string, and when set the parameter is set as expected

Added some integration tests that excercise some of the new options

## Discussion

~~I'd appreciate feedback on the implementation for ["excluding" labels](https://help.github.com/articles/searching-issues/#search-by-the-labels-on-an-issue).  In the API docs, this is done by using notation `-label:labelname` as opposed to `label:labelname` to include a label.  At first I thought we could just implement this by having any label prepended with a hyphen be deemed to be an "exclusion" label, however it turns out that it's actually valid to have labels that start with a hyphen.  So I implemented this as a separate list property `SearchIssuesRequest.NotLabels`~~

~~It feels a bit ugly so if someone has a better way, let me know~~

~~**Update**: Now renamed to `SearchIssuesRequest.ExcludeLabels`~~
See https://github.com/octokit/octokit.net/pull/1228#issuecomment-203965890 for how the "exclusion" field implementation is shaping up...

Similarly, the ["missing metadata"](https://help.github.com/articles/searching-issues/#search-by-missing-metadata-on-an-issue-or-pull-request) search option, ive called the class property `No` to match the query property in API docs `no`, and gone with `IssueNoMetadataQualifier` as the enum name.  Open to better suggestions...

### CODE REVIEW

**Comment by M-Zuber** [comment]
A few things off the top of my head:
- Have a test that uses a label starting with a hyphen and exclude it
  - Codifies the reasoning behind having that separate list
- Can a username also start with a `-`?
  - If so, excluding issues by a given user would also necessitate a similar construct

**Comment by ryangribble** [comment]
I dont see anything in the API docs about being able to "exclude" any of the other filter types...

The only thing mentioned is labels (`-label:labelname` vs `label:labelname`) and keywords (specifying `NOT fish` instead of `fish` as the search term).

> Have a test that uses a label starting with a hyphen and exclude it

With the way it's been implemented the label name doesn't really matter.   I have a unit test that checks when 2 labels are specified in `NotLabel` we get the expected `-label:label1` `-label:label2` arguments.  Also I added an integration test that searches octokit.net repo for "up-for-grabs" labelled issues and a 2nd search for NOT "up-for-grabs" labelled issues, then checks to make sure the lists are unique

**Comment by M-Zuber** [comment]
image

It works on (almost) anything

>>Have a test that uses a label starting with a hyphen and exclude it

>With the way it's been implemented the label name doesn't really matter. I have a unit test that checks when 2 labels are specified in NotLabel we get the expected -label:label1 -label:label2 arguments. Also I added an integration test that searches octokit.net repo for "up-for-grabs" labelled issues and a 2nd search for NOT "up-for-grabs" labelled issues, then checks to make sure the lists are unique

What I was trying to say was have a test that looks for a label that starts with a `-` so as to show that it is legal for a label to start with that char, and it will not be treated as negation if such a label exists.
Not showing that `NotLabel` will exclude it, but that Label/Term will include it

**Comment by shiftkey** [comment]
@M-Zuber be careful, the repository issue search on the website is likely to be a bit different to what's available through the API. Someone asked me a while ago about this and I'm pretty sure `-label` doesn't work through the API when looking at repository issues, for example.

@ryangribble regarding your question about the properties, an alternative might be `ExcludeLabels`. Personally, I don't think that users should need to know or care that you need to do a `-` prefix, so this is a good direction to head.

**Comment by M-Zuber** [comment]
- :embarrassed: guess I just assumed the api was being dogfooded. 
- as someone who wrote (albeit a simple one) a query language, hiding syntax behind human readable properties is A Very Good Thing™

**Comment by ryangribble** [comment]
- rebased on latest master

- renamed `NotLabels` to `ExcludeLabels`

- I also tidied up an inconsistency that was irking me.  We refer to pull requests as `PullRequest` pretty much everywhere in octokit.net - except the `IssueTypeQualifier.PR` enum value.  So to maintain consistent naming and terminology, Ive added `IssueTypeQualifier.PullRequest` to that enum and flagged the `PR` value as `[Obsolete]` so it can be removed sometime down the track :smile: 

>  guess I just assumed the api was being dogfooded.
It could well be that the API accepts more options than the doco says, but for now Im thinking of just implementing what the doc says (where `labels` are the only things that can be excluded).  

Although Im thinking perhaps for future proofing (If it does turn out that lots more fields can be negated), I could perhaps move to having a new class to hold all the exclusions (and thus refer to them as normal names like `Label`)

``` csharp
new SearchIssuesRequest("fish")
{
  Type = IssueTypeQualifier.PullRequest,
  Is = IssueIsQualifier.Merged,
  Exclusions = new SearchIssuesExclusions
  {
      Labels = new[] { "label1", "label2" },
      Author = "M-Zuber" // Docs dont say this is supported but this is where more exclude fields could go if/when added
  }
}
```

Thoughts?

**Comment by M-Zuber** [comment]
>Although Im thinking perhaps for future proofing (If it does turn out that lots more fields can be negated or even the fact that it's likely in the future more fields would be able to be), I could perhaps move to having a new class to hold all the exclusions (and thus refer to them as normal names like Label).

That is pretty much what I was trying to say with this :arrow_down:, and your example implementation is nicer :smile: 
>Edit:
Real bran-fart this time: If a number of filterable items can start with a - (meaning to actually filter you'd need to do a --), maybe make a .Not property that contains a nested SearchIssuesRequest? (I know crazy idea but you only live once or something)

**Comment by M-Zuber** [comment]
> It could well be that the API accepts more options than the doco says, but for now Im thinking of just implementing what the doc says (where labels are the only things that can be excluded).

Sounds good to me, I just wanted to bring up such a possibility

**Comment by ryangribble** [comment]
OK so ive just pushed up a change that has moved the exclusions fields out into a subclass, as discussed above.  So they are now accessed via `SearchIssuesRequest.Exclusions`

eg
``` csharp
    // Search for PullRequests that dont have "skip-release-notes" label
    var request = new SearchIssuesRequest();
    request.Type = IssueTypeQualifier.PullRequest;
    request.Repos.Add("octokit", "octokit.net");
    request.Exclusions = new SearchIssuesRequestExclusions
    {
        Labels = new[] { "skip-release-notes" }
    };

    var pullRequests = await _gitHubClient.Search.SearchIssues(request);
```

Regarding which fields are supported, I did some poking at the API and determined that it actually does support quite a few fields in "exclusion" syntax, in addition to the mentioned "`label`" field.

So the following fields have been implemented in the Exceptions parameters:
- `Author, Assignee, Mentions, Commenter, Involves, State, Labels, Language, Status, Head, Base`
These were all tested and work correctly - integration tests are added to assert this.
Although not mentioned in the docs (except `Labels`), these do seem useful (ie to search for items where Assignee is not me, or shiftkey is not mentioned etc)

The following fields were not implemented for exclusion for listed reasons:
- `Type,In,No,Is`
These failed integration tests, API doesnt support negating these

- `Created,Updated,Merged,Closed,Comments`
These date/range fields already support greater/less than and "between" syntax, so having an exclusion option doesnt really make sense so I didnt attempt to implement/test

- `Team`
I didnt get time to find a suitable repo that has teams being @mentioned in issues, to run integration tests against

- `User`
Not tested.  
Does it make sense to do a search for something in all repos EXCEPT those owned by a user or organisation?

- `Repos`
Not tested.  
Does it make sense to do a search in all repos EXCEPT some specified?


Thoughts? 
Will definitely need some guidance from @Haacked and @shiftkey on this
On the one hand it seems prudent to only implement the mentioned `label` field for exclusion and "dangerous" to implement something that isnt listed in the API docs.  But on the other, being able to exclude by author, target branch etc have real life beneficial use cases, and the API clearly does support them (and the integration tests will detect if they ever cease to work)...

**Comment by M-Zuber** [comment]
> Created,Updated,Merged,Closed,Comments
These date/range fields already support greater/less than and "between" syntax, so having an exclusion option doesnt really make sense so I didnt attempt to implement/test

I would say to yes implement (assuming support) as some users may find it easier conceptually to use `.Exclusions` (BTW great job on naming)
Might be open to doing a PR into this PR for those fields.
>User
Not tested.
Does it make sense to do a search for something in all repos EXCEPT those owned by a user or organisation?
Repos
Not tested.
Does it make sense to do a search in all repos EXCEPT some specified?

I vote for yes as I have done both in the past.

One example that covers both is looking for something .Net Core related - but not in anything from Microsoft & Microsoft related repositories.

**Comment by ryangribble** [comment]
Cool well yeah i didnt actually even test/probe those options to see whether api accepts excluding them or not... i will look into it

**Comment by ryangribble** [comment]
Hey @shiftkey just a reminder that I'm waiting on some input from your good self before I take any next steps :grinning: 

I know you've been on the road etc, hoping that since you've assigned this to yourself it is somewhere on your list to get to :crystal_ball:

**Comment by shiftkey** [comment]
@ryangribble I haven't forgotten, just been digging myself out from under work since I got back to Oz last week and needed a bit of mundane work to take a break. I'll pick this up tomorrow.

**Comment by shiftkey** [comment]
@ryangribble looking over the recent questions:

> The following fields were not implemented for exclusion for listed reasons:
> - `Type,In,No,Is`
> These failed integration tests, API doesnt support negating these

Works for me.

> - `Created,Updated,Merged,Closed,Comments`
> These date/range fields already support greater/less than and "between" syntax, so having an exclusion option doesnt really make sense so I didnt attempt to implement/test

Yep, I like this too.

> - `Team`
> I didnt get time to find a suitable repo that has teams being @mentioned in issues, to run integration tests against

This is fine. We'll cross that bridge if we have to.

> - `User`
> Not tested.  
> Does it make sense to do a search for something in all repos EXCEPT those owned by a user or organisation?

Happy to kick this can down the road until someone comes looking for it.

> - `Repos`
> Not tested.  
> Does it make sense to do a search in all repos EXCEPT some specified?

I guess if you wanted to offload more of the search work onto the API. But again, really edge case. :boot: this for now.

File: Octokit.Tests.Integration/Clients/SearchClientTests.cs [hunk 1]
```diff
@@ -126,4 +126,392 @@ public SearchClientTests()
         Assert.NotEmpty(closedIssues);
         Assert.NotEmpty(openedIssues);
     }
+
+    [Fact]
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/SearchClientTests.cs]: I think these should all be `[IntegrationTest]` to ensure they only run when you have an integration test account setup.

File: Octokit.Tests.Integration/Clients/SearchClientTests.cs [hunk 2]
```diff
@@ -126,4 +126,392 @@ public SearchClientTests()
...
+            Assignee = assignee
+        };
+
+        var otherIssues = await _gitHubClient.Search.SearchIssues(excludeRequest);
+
+        // Ensure we found issues
+        Assert.NotEmpty(otherIssues.Items);
+
+        // Ensure no items from the first search are in the results for the second
+        Assert.DoesNotContain(issues.Items, x1 => otherIssues.Items.Any(x2 => x2.Id == x1.Id));
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/SearchClientTests.cs]: Trying to think of a more succinct way to describe this, but coming up empty 👌

File: Octokit.Tests/Models/SearchIssuesRequestTests.cs [hunk 3]
```diff
@@ -25,5 +27,156 @@ public void SortNotSpecifiedByDefault()
...
+
+                // Ensure the specified parameter now exists
+                Assert.True(request.MergedQualifiers().Count(x => x.Contains(property.Item1)) == 1);
+            }
+        }
+
+        [Fact]
+        public void HandlesDateRangeAttributesCorrectly()
+        {
+            var dateProperties = new List<Tuple<string, Action<SearchIssuesRequest, DateRange>>>()
```
Comment (shiftkey) [inline Octokit.Tests/Models/SearchIssuesRequestTests.cs]: Semantics, but perhaps `Dictionary<string,Action<SearchIssuesRequest, DateRange>>` instead of crafting tuples?

**Comment by shiftkey** [comment]
Changes look good, integration tests are working fine on my end.

I'm happy with the direction, happy to give it another look now that I finally got around to answering your questions...

**Comment by ryangribble** [comment]
Thanks @shiftkey ill tidy up those few things raised.

There was also I guess 2 other things I just wanted a specific 👍  from you on:

> I also tidied up an inconsistency that was irking me. We refer to pull requests as PullRequest pretty much everywhere in octokit.net - except the `IssueTypeQualifier.PR` enum value. So to maintain consistent naming and terminology, Ive added `IssueTypeQualifier.PullRequest` to that enum and flagged the `PR` value as `[Obsolete]` so it can be removed sometime down the track :smile:

and

> Similarly, the "missing metadata" search option, ive called the class property `No` to match the query property in API docs `no`, and gone with `IssueNoMetadataQualifier` as the enum name. Open to better suggestions...

**Comment by shiftkey** [comment]
:thumbsup: to `PR` -> `PullRequest` - and :heart: for doing the obsoletion already
:thumbsup: for `No`

File: Octokit.Tests.Integration/Clients/SearchClientTests.cs (hunk 1 above)
Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/SearchClientTests.cs]: Nice spot, thanks

File: Octokit.Tests/Models/SearchIssuesRequestTests.cs (hunk 3 above)
Comment (ryangribble) [inline Octokit.Tests/Models/SearchIssuesRequestTests.cs]: Poor old Tuples everyone hates them!  Yep good idea not sure how I ended up with `<List<Tuple<` instead of just a `Dictionary` hahah :grinning: 

These are now tidied up

File: Octokit.Tests.Integration/Clients/SearchClientTests.cs [hunk 4]
```diff
@@ -126,4 +126,392 @@ public SearchClientTests()
         Assert.NotEmpty(closedIssues);
         Assert.NotEmpty(openedIssues);
     }
+
+    [IntegrationTest]
+    public async Task SearchForMergedPullRequests()
+    {
+        var allRequest = new SearchIssuesRequest();
+        allRequest.Repos.Add("octokit", "octokit.net");
+        allRequest.Type = IssueTypeQualifier.PR;
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/SearchClientTests.cs]: :lipstick: rename this to `.PullRequest`

Comment (ryangribble) [inline Octokit.Tests.Integration/Clients/SearchClientTests.cs]: Lol, nice pickup!

**Comment by ryangribble** [comment]
All feedback addressed, merged latest master and bashed Travis until it finally got a green run!

**Comment by shiftkey** [comment]
@ryangribble +:100: nice work!
